### DISCUSSION TITLE
Text watermark background color

### DISCUSSION THREAD

**Discussion by dobras** [SixLabors/ImageSharp#1649] [Q&A] [answered]
Hi everyone! Im trying the library to find out a good library for our project and so far this looks great, but i came to a problem. Is there a way to add black background to watermark text?

I need this to be the result, but i am failing in adding the background.
![image](https://user-images.githubusercontent.com/17703085/120498942-edc1c100-c3bf-11eb-8835-3a71302d6b0f.png)

This i my result so far:
![image](https://user-images.githubusercontent.com/17703085/120502537-cd473600-c3c2-11eb-89ee-2f4f9731d860.png)

And this is the sourcecode:
```
        public async Task AddWatermark(string watermarkText)
        {
            string input = Path.Combine(this.path, this.inputFile);
            string output = Path.Combine(this.path, this.outputFile);

            FontCollection fontCollection = new FontCollection();
            FontFamily fontFamily = fontCollection.Install(@"C:/Windows/fonts/consola.ttf");
            Font font = fontFamily.CreateFont(9);
           
            using Image image = await Image.LoadAsync(input);
            image.Mutate(context =>
            {
                context.ApplyScalingWaterMark(watermarkText, Color.White, font, 6);
            });
            await image.SaveAsync(output);
        }

        public static IImageProcessingContext ApplyScalingWaterMark(this IImageProcessingContext processingContext, string text, 
        Color color, Font font, int scalingFactor = 1)
        {
        	  // measure the text size
            FontRectangle textSize = TextMeasurer.Measure(text, new RendererOptions(font));
            
            //find out how much we need to scale the text to fill the space (up or down)
            Font scaledFont = new Font(font, scalingFactor * font.Size);

            PointF bottomRight = new PointF(imageSize.Width - textSize.Width, imageSize.Height - textSize.Height);

            DrawingOptions options = new DrawingOptions();
            options.TextOptions = new TextOptions
            {
                HorizontalAlignment = HorizontalAlignment.Right,
                VerticalAlignment = VerticalAlignment.Bottom
            };

            processingContext.BackgroundColor(Color.Black);
            return processingContext.DrawText(options, text, scaledFont, color, bottomRight);
        }
```

Sorry for the different pictures but they are not importnant :D. 
Many Thanks for your time.

**Comment by JimBobSquarePants** [comment, marked as answer]
```c#
processingContext.BackgroundColor(Color.Black);
return processingContext.DrawText(options, text, scaledFont, color, bottomRight);
```

You're asking context here to set the background color of the image not of the rectangle where the text shall be drawn. 
You should [draw a rectangle at the desired location](https://docs.sixlabors.com/api/ImageSharp.Drawing/SixLabors.ImageSharp.Drawing.Processing.FillRectangleExtensions.html), then draw the text.

**Reply by dobras** [reply]
Thanks for the super quick reply! Yeah i forgot this line there :D. But basicaly i should not think about it as setting background color of a text but drawing a background of the text and then drawing text on the top of it.

**Reply by JimBobSquarePants** [reply]
Yep. Exactly.

**Reply by dobras** [reply]
Also i noticed that the images that i added watermark to are considerably larger than those with normal operation (resize, crop ....)
What am i doing wrong pleeaaase :D ?
I can also provide the project itself.

![image](https://user-images.githubusercontent.com/17703085/120799527-250aac00-c53f-11eb-8d67-2810ae674c14.png)

and again here is my code which changed only slightly from last time.
```
        public async Task AddWatermark(string watermarkText, int scalingFactor = 1)
        {
            string input = Path.Combine(this.path, this.inputFile);
            string output = Path.Combine(this.path, this.outputFile);

            FontCollection fontCollection = new FontCollection();
            FontFamily fontFamily = fontCollection.Install(@"C:/Windows/fonts/consola.ttf");
            Font font = fontFamily.CreateFont(9);
            Font scaledFont = new Font(font, scalingFactor * font.Size);

            using Image image = await Image.LoadAsync(input);
            image.Mutate(context => { context.ApplyScalingWaterMark(watermarkText, Color.White, scaledFont); });
            await image.SaveAsync(output);
        }

public static class ImageProcessingContextExtensions
{
        public static IImageProcessingContext ApplyScalingWaterMark(this IImageProcessingContext processingContext, string 
        text, Color color, Font font)
        {
            Size imageSize = processingContext.GetCurrentSize();

            // measure the text size
            FontRectangle textDimensions = TextMeasurer.Measure(text, new RendererOptions(font));
            SizeF textSize = new SizeF(textDimensions.Width, textDimensions.Height);

            PointF bottomRight = new PointF(imageSize.Width - textSize.Width, imageSize.Height - textSize.Height);

            SizeF textBackgroundSize = new SizeF(textSize);
            RectangleF textBackground = new RectangleF(bottomRight, textBackgroundSize);

            return processingContext
                .Fill(Color.Black, textBackground)
                .DrawText(text, font, color, bottomRight);
        }
}

```

**Reply by dobras** [reply]
Should i write the followup question in a new thread and mark this one solved? The followup question is kinda connected to the previous one but i understand if you feel like not.

**Reply by antonfirsov** [reply]
It's not fully clear what are you comparing to what. (Original image VS output with watermark *or* output with VS without watermark?)

In general, you need to tweak the encoder parameters manually to reduce the output size:
https://docs.sixlabors.com/articles/imagesharp/imageformats.html#working-with-encoders

**Reply by dobras** [reply]
Thank you for the repply. Sorry i see, I was comparing original (guid.jpg) with both outputs (output_watermark.png and output_addwatermark_fromStream.png) and original has 1.4mb and after adding watermark, in the better case result was 6.3mb and the other was 10.7mb which is way to much for just added rectangle with text. Ill check the link.

**Reply by dobras** [reply]
Ok so have been trying to use the specific encoders/decoder but to no success. Dont know what im doing wrong?
What makes the image to become 4-6 times larger than original?
I found out that with this code - when im loading jpg and savin png that creates the big increase, but when i load jpg and save jpg, then everything works correctly. 
That makes me wonder, what is the correct process to change the format of the image in the process if not defining encoders and decoders?
As i said before, i can provide the source codes no problem. 

```
public async void AddWatermark(string watermarkText, int scalingFactor = 1)
        {
            string input = Path.Combine(this.path, this.inputFile);
            string output = Path.Combine(this.path, this.outputFile);

            FontCollection fontCollection = new FontCollection();
            FontFamily fontFamily = fontCollection.Install(@"C:/Windows/fonts/consola.ttf");
            Font font = fontFamily.CreateFont(9);
            Font scaledFont = new Font(font, scalingFactor * font.Size);

            using Image image = await Image.LoadAsync(input, new JpegDecoder());
            
            image.Mutate(context => { context.ApplyScalingWaterMark(watermarkText, Color.White, scaledFont); });
            await image.SaveAsync(output, new PngEncoder());
        }
```

**Reply by antonfirsov** [reply]
> when im loading jpg and savin png that creates the big increase

PNG is a lossless format, it's not unexpected that files get much bigger when resaved to PNG, basically the same will happen when you resave a Jpeg as Png with a graphics editor software like Paint.

**Reply by dobras** [reply]
Oh yes you are right! Im sorry, its my lack of knowledge of the images. Thank you very much!
